### PR TITLE
disguise api

### PR DESCRIPTION
**Pull request by yannicklamprecht**
relates to https://github.com/PaperMC/Paper/discussions/10584

Adds a simple disguise api that also makes sure that invalid entity data, not belonging to the type that is faked, is filtered.
It also adds a server generator module to reduce maintenance of the entity data filter.

tldr; The API provides a way to disguise any entity as another or a player with having the possibility to interact with the entity like with any others. The big advantage is that you've the whole vanilla logic and Mob Goal and Pathfinding included without hassle.

Testing the changes can be done using the following snippet:
https://gist.github.com/yannicklamprecht/639acba127a54f5617e1281f099b7aff

TODO:

- [x] **Add JavaDoc**
- [x] remove test plugin code
- [x] comment paper-server-generator
- [x] ~~figure out how to get teams working with real teams, client side ones worked in my fork~~ Non issue just use the `Entity#getName()` instead of the `Entity#getScoreboardEntryName()` 
- [x] remove player profile if player disguise on ServerEntiry#removePairing(player) to avoid client disconnects when moving away and approaching again. see https://github.com/PaperMC/Paper/pull/10478#issuecomment-2105299573
- [x] wait until dev/1.21 is merged into master
- [x] When player is disguised as entity his attributes are propagated to the client, which isn't supported by certain entities
- [x] When player is disguised as entity that has no ability to interact/build, the server should not propagate the players action to the rest of the players. Otherwise the other player gets kicked.
- [x] When changing disguise while in visible range from e.g. Boat to a Living entity disguise, attributes will be sent again. -> on disguise change => auto hide and show entity to players.

Effective loc:
- 600 loc in patches
- 3k+ loc generated filter classe
- 400 loc generator code copied from api-generator
- 100 loc for generating the filter class.

### CODE REVIEW

File: patches/server/1047-add-disguise-api.patch
Comment (yannicklamprecht) [inline patches/server/1047-add-disguise-api.patch]: unindent.

File: patches/api/0473-add-disguise-api.patch
Comment (yannicklamprecht) [inline patches/api/0473-add-disguise-api.patch]: Fix setter.
```suggestion
+    void setDisguiseData(@NotNull io.papermc.paper.disguise.DisguiseData disguiseData);
```

File: patches/api/0473-add-disguise-api.patch
Comment (yannicklamprecht) [inline patches/api/0473-add-disguise-api.patch]: Just delegate.
```suggestion
+        return reset();
```

File: paper-server-generator.settings.gradle.kts
Comment (yannicklamprecht) [inline paper-server-generator.settings.gradle.kts]:
```suggestion
 // include(":paper-server-generator")
```

File: patches/server/1047-add-disguise-api.patch
Comment (yannicklamprecht) [inline patches/server/1047-add-disguise-api.patch]: Move creation to methods to avoid duplication.

File: patches/api/0475-add-disguise-api.patch
Comment (kashike) [inline patches/api/0475-add-disguise-api.patch]: Should be formatted like so:
```suggestion
+        @org.jetbrains.annotations.NotNull Builder withCape();
```

File: patches/api/0475-add-disguise-api.patch
Comment (kashike) [inline patches/api/0475-add-disguise-api.patch]: These should all accept a `boolean` argument.

**Review by kashike — CHANGES_REQUESTED** [review]

File: patches/server/1047-add-disguise-api.patch
Comment (MiniDigger) [inline patches/server/1047-add-disguise-api.patch]: some of the classes here are imported and some use the fully qualified name (even if they are imported)
just use imports for everything, its a class we add so no risk of dum conflicts

File: patches/server/1047-add-disguise-api.patch
Comment (MiniDigger) [inline patches/server/1047-add-disguise-api.patch]: I would keep the empty line here and below to have nicer looking diff

File: patches/server/1047-add-disguise-api.patch
Comment (MiniDigger) [inline patches/server/1047-add-disguise-api.patch]: MAJOR nit pick but if you name the arugment packet the method looks a whole lot nicer, lol

**Comment by yannicklamprecht** [comment]
Tested with:
https://gist.github.com/yannicklamprecht/639acba127a54f5617e1281f099b7aff

Only open part is the discussion for a common generator module

**Comment by yannicklamprecht** [comment]
We should probably move the common generator classes into a separate module to reduce duplications.

**Comment by MiniDigger** [comment]
this is complicated enough that we prolly want a lil docs page with an example, but that can wait till after its merged and we know the final api design

**Comment by yannicklamprecht** [comment]
> this is complicated enough that we prolly want a lil docs page with an example, but that can wait till after its merged and we know the final api design

Thought of adding a Java 18+ snippet to the javadoc but a docs page will be better.

**Comment by MiniDigger** [comment]
I meant the api itself

**Comment by yannicklamprecht** [comment]
Addressed issue with experience orbs by filtering them out. Mentioned that exclusion in the JavaDoc.

Additionally found an issue with ElderGuardian when disguising them as a Player on the client side. That happens only when the ElderGuardian is in the water. First thought is the ElderGuardian stream sucking health from other entities. To be sure I need to deobf the client stacktrace.

```java
Failed to handle packet adm@395c0068
java.lang.ClassCastException: class gct cannot be cast to class ckb (gct and ckb are in unnamed module of loader 'app')
	at fxy.a(SourceFile:1068)
	at adm.a(SourceFile:41)
	at adm.a(SourceFile:13)
	at zz.a(SourceFile:25)
	at bpm.d(SourceFile:162)
	at bpq.d(SourceFile:23)
	at bpm.A(SourceFile:136)
	at bpm.bA(SourceFile:121)
	at ffh.c(SourceFile:1275)
	at ffh.f(SourceFile:888)
	at net.minecraft.client.main.Main.main(SourceFile:265)
```

**Comment by kennytv** [comment]
That's ClientPacketListener handleEntityEvent

**Comment by yannicklamprecht** [comment]
> That's ClientPacketListener handleEntityEvent, Player can't be cast to Guardian

Found the reason why. I basically moved away from the entity, the chunk or at least the entities got unloaded.
Due to the state that the disguise data is not persisted on the entity, when loading the entity again it published itself to the client without the disguise info. The client knows the entity uuid only as the disguised type but gets the original one.

I updated the test plugin snippet to reflect the needed change - cache cleaning on death excluded.

It would be way better to persist the disguise data in NBT to avoid manual caching per plugin.

**Comment by yannicklamprecht** [comment]
Addressed the last known issue by removing the profile info for a player disguised entity on ServerEntity#removePairing(player).
https://github.com/PaperMC/Paper/pull/10478/files#diff-1b00ff7ef0678fae6f893a77a2ed93bece7dbbd375a03115590b25304423ab1fR161-R170

https://github.com/PaperMC/Paper/pull/10478/files#diff-1b00ff7ef0678fae6f893a77a2ed93bece7dbbd375a03115590b25304423ab1fR252

**Comment by yannicklamprecht** [comment]
Update to 1.21

**Comment by XxGoldenbluexX** [comment]
good job!

File: patches/server/1057-add-disguise-api.patch
Comment (Lulu13022002) [inline patches/server/1057-add-disguise-api.patch]: This will always return Entity.class, so the generated content is almost useless and all the metadata are skipped.
Generally i don't think the new module is really needed and this can be done with regular reflection like https://pastes.dev/FNvwYGnwgS

Comment (yannicklamprecht) [reply patches/server/1057-add-disguise-api.patch]: I'm against using reflection for that. It can hide potential bugs occuring during runtime. When generating everything during update time we can find issues way faster.

I addressed your concern regarding the base class issue by generating a mapping.

https://github.com/PaperMC/Paper/pull/10478/files#diff-a2c5f4e52ea4c2f48ddb4719a5c436f964a8e49beb30c3c90f099d4b8f94e18a

https://github.com/PaperMC/Paper/pull/10478/files#diff-e150aa8e1c1a506433452cd2d30284e9d7bba7a54f13d8c0faef70e715d816a4R150

File: patches/server/1061-add-disguise-api.patch
Comment (yannicklamprecht) [inline patches/server/1061-add-disguise-api.patch]: The following lines ensure that a force refresh is triggered for all players that can see the entity.

**Comment by MiniDigger** [comment]
Hey @yannicklamprecht, since I see you rebasing this PR all the time: you really don't need to do that for this particular one.
While we do like this API in principle, it opened up a lot of discussion internally and we are not really sure yet on *how* we want to land it. We do have an idea on how it could look like post hard fork, but for now you don't need to waste time updating it, unless you need it for your own uses of course.
Thanks for sticking with us!

**Comment by yannicklamprecht** [comment]
> Hey @yannicklamprecht, since I see you rebasing this PR all the time: you really don't need to do that for this particular one. While we do like this API in principle, it opened up a lot of discussion internally and we are not really sure yet on _how_ we want to land it. We do have an idea on how it could look like post hard fork, but for now you don't need to waste time updating it, unless you need it for your own uses of course. Thanks for sticking with us!



1. It's a no brainer to update the pr.
2. I need the feature in my own fork and I'm to lazy to workaround the fact that code generation for forks isn't working that easy after a code gen change somewhere this year.

**Comment by yannicklamprecht** [comment]
Post hardfork MR: https://github.com/PaperMC/Paper/pull/11793

**Pull request closed without merging**
